src/primval.rs

```diff
@@ -69,15 +69,26 @@ pub fn binary_op<'tcx>(bin_op: mir::BinOp, left: PrimVal, right: PrimVal) -> Eva
     match bin_op {
         // can have rhs with a different numeric type
         Shl | Shr => {
+            let mask_bits = match left {
```

Comment:
@oli-obk Can you add a comment explaining the masking? It's a bit non-obvious.

@eddyb Is there a source in rustc or the docs we can refer to for this?